File: experiments/pm-synth/src/audio/modules/interface/strategies/PlasmaStrategy.js
// PlasmaStrategy.js
// Plasma interface: Electromagnetic waveguide with nonlinear dispersion

import { InterfaceStrategy } from '../InterfaceStrategy.js';
import { AmplitudeTracker } from '../utils/EnergyTracker.js';
import { cubicWaveshaper } from '../utils/NonlinearityLib.js';

export class PlasmaStrategy extends InterfaceStrategy {
    constructor(sampleRate) {
        super(sampleRate);

        // Amplitude tracking for nonlinear dispersion
        this.ampTracker = new AmplitudeTracker(0.001, sampleRate);

        // Phase accumulator for self-focusing effects
        this.phase = 0;

        // Previous samples for allpass-style dispersion
        this.x1 = 0;
        this.y1 = 0;
    }

    /**
     * Process one sample through plasma interface
     * Amplitude-dependent phase modulation (self-focusing)
     * @param {number} input - Input sample
     * @returns {number} Processed output
     */
    process(input) {
        // Track amplitude for nonlinear effects
        const amplitude = this.ampTracker.process(input);

        // Self-focusing: high amplitude -> faster propagation
        const beta = this.intensity * 0.3;
        const phaseMod = 1 + beta * amplitude;

        this.phase += 0.1 * phaseMod;
        if (this.phase > Math.PI * 2) {
            this.phase -= Math.PI * 2;
        }

        // Amplitude-to-frequency conversion
        const freqMod = Math.sin(this.phase) * amplitude * this.intensity * 0.5;

        // Dispersive allpass filter with amplitude-dependent coefficient
        const allpassCoeff = 0.3 + amplitude * this.intensity * 0.4;
        const dispersed = allpassCoeff * input + this.x1 - allpassCoeff * this.y1;

        this.x1 = input;
        this.y1 = dispersed;

        // Add frequency modulation component
        const output = dispersed + freqMod;

        // Nonlinear harmonic generation at high intensities
        if (this.intensity > 0.5) {
            const nonlinear = cubicWaveshaper(output, (this.intensity - 0.5) * 0.4);
            return Math.max(-1, Math.min(1, nonlinear));
        }

        return Math.max(-1, Math.min(1, output));
    }

    /**
     * Reset state
     */
    reset() {
        this.ampTracker.reset();
        this.phase = 0;
        this.x1 = 0;
        this.y1 = 0;
    }

    /**
     * Called on note-on trigger
     */
    onNoteOn() {
        this.reset();
    }
}

File: gtk-synth/src/audio/modules/strategies/plasma_strategy.c
// plasma_strategy.c
// Plasma interface: Electromagnetic waveguide with nonlinear dispersion
// Translated from reference/modules/interface/strategies/PlasmaStrategy.js

#include "interface_strategy.h"
#include "dsp_utils.h"
#include <stdlib.h>
#include <math.h>

typedef struct {
    AmplitudeTracker amp_tracker;
    float phase;
    float x1;
    float y1;
} PlasmaImpl;

static float plasma_process(InterfaceStrategy* self, float input) {
    PlasmaImpl* impl = (PlasmaImpl*)self->impl_data;

    // Track amplitude for nonlinear effects
    const float amplitude = amplitude_tracker_process(&impl->amp_tracker, input);

    // Self-focusing: high amplitude -> faster propagation
    const float beta = self->intensity * 0.3f;
    const float phase_mod = 1.0f + beta * amplitude;

    impl->phase += 0.1f * phase_mod;
    if (impl->phase > M_PI * 2.0f) {
        impl->phase -= M_PI * 2.0f;
    }

    // Amplitude-to-frequency conversion
    const float freq_mod = sinf(impl->phase) * amplitude * self->intensity * 0.5f;

    // Dispersive allpass filter with amplitude-dependent coefficient
    const float allpass_coeff = 0.3f + amplitude * self->intensity * 0.4f;
    const float dispersed = allpass_coeff * input + impl->x1 - allpass_coeff * impl->y1;

    impl->x1 = input;
    impl->y1 = dispersed;

    // Add frequency modulation component
    float output = dispersed + freq_mod;

    // Nonlinear harmonic generation at high intensities
    if (self->intensity > 0.5f) {
        const float nonlinear = cubic_waveshaper(output, (self->intensity - 0.5f) * 0.4f);
        return fmaxf(-1.0f, fminf(1.0f, nonlinear));
    }

    return fmaxf(-1.0f, fminf(1.0f, output));
}

static void plasma_reset(InterfaceStrategy* self) {
    PlasmaImpl* impl = (PlasmaImpl*)self->impl_data;
    amplitude_tracker_reset(&impl->amp_tracker);
    impl->phase = 0.0f;
    impl->x1 = 0.0f;
    impl->y1 = 0.0f;
}

static void plasma_set_intensity(InterfaceStrategy* self, float intensity) {
    self->intensity = intensity;
}

static void plasma_set_gate(InterfaceStrategy* self, bool gate) {
    if (gate) {
        plasma_reset(self);
    }
}

static void plasma_destroy(InterfaceStrategy* self) {
    if (self) {
        free(self->impl_data);
        free(self);
    }
}

static InterfaceStrategyVTable plasma_vtable = {
    .process = plasma_process,
    .reset = plasma_reset,
    .set_intensity = plasma_set_intensity,
    .set_gate = plasma_set_gate,
    .destroy = plasma_destroy
};

InterfaceStrategy* plasma_strategy_create(float sample_rate) {
    (void)sample_rate;  // Unused
    InterfaceStrategy* strategy = (InterfaceStrategy*)calloc(1, sizeof(InterfaceStrategy));
    if (!strategy) return NULL;

    PlasmaImpl* impl = (PlasmaImpl*)calloc(1, sizeof(PlasmaImpl));
    if (!impl) {
        free(strategy);
        return NULL;
    }

    amplitude_tracker_init(&impl->amp_tracker, 0.001f, 44100.0f);
    impl->phase = 0.0f;
    impl->x1 = 0.0f;
    impl->y1 = 0.0f;

    strategy->vtable = &plasma_vtable;
    strategy->impl_data = impl;
    strategy->intensity = 0.5f;

    return strategy;
}
